impl/src/main/java/org/ehcache/impl/internal/classes/ClassInstanceProvider.java

```diff
@@ -138,9 +146,10 @@ protected void releaseInstance(T instance) throws IOException {
       throw new IllegalArgumentException("Given instance of " + instance.getClass().getName() + " is not managed by this provider");
     }
 
-    if(instance instanceof Closeable) {
+    if(instantiated.contains(instance) && instance instanceof Closeable) {
```

Comment:
This can be simplified by doing `if (instantiated.remove(instance)) && instance instance Closeable)` and thus dropping the remove after the block.